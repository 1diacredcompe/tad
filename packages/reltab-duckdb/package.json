{
  "name": "reltab-duckdb",
  "version": "0.10.0",
  "description": "sqlite driver for reltab",
  "main": "dist/reltab-duckdb.js",
  "types": "dist/reltab-duckdb.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc -p tsconfig-build.json",
    "test": "jest --config jest.config.json --no-cache"
  },
  "keywords": [
    "relational",
    "sql",
    "database",
    "sqlite"
  ],
  "author": "Antony Courtney <antony@antonycourtney.com>",
  "license": "MIT",
  "dependencies": {
    "byline": "^5.0.0",
    "csv-sniffer": "^0.1.1",
    "fast-csv": "^4.3.6",
    "gauge": "^3.0.1",
    "loglevel": "^1.7.1",
    "through": "^2.3.8",
    "typed-promisify": "^0.4.0",
    "duckdb": "^0.3.1",
    "node-duckdb": "^0.0.76",
    "pretty-hrtime": "^1.0.3"
  },
  "devDependencies": {
    "@types/pretty-hrtime": "^1.0.1",
    "@types/through": "0.0.30",
    "@types/byline": "^4.2.33",
    "@types/jest": "^26.0.24",
    "@types/node": "^16.4.13",
    "jest": "^27.0.6",
    "mkdirp": "^1.0.4",
    "prettier": "^2.3.2",
    "ts-jest": "^27.0.4",
    "typescript": "^3.9.4"
  },
  "peerDependencies": {
    "reltab": "^0.10.0"
  }
}
