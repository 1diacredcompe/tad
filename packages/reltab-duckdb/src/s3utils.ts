/**
 * Support routines for using s3 URLs with DuckDb.
 */
import { Connection, DuckDB } from "ac-node-duckdb";
import * as log from "loglevel";

/**
 * Initialize S3 variables on a DuckDb connection from env vars
 * It's a bit unfortunate to do this every time we create a DuckDb
 * connection, but the s3fs extension depends on these vars being
 * set. Hopefully minimal overhead, and we skip if env vars
 * not set.
 */
const AWS_DEFAULT_REGION = "us-west-1";

export async function initS3(dbConn: Connection) {
  const aws_access_key_id = process.env.AWS_ACCESS_KEY_ID;
  if (aws_access_key_id && aws_access_key_id.length > 0) {
    log.debug("AWS_ACCESS_KEY_ID env var found, initializing S3 vars");
    const s3_region = process.env.AWS_REGION ?? AWS_DEFAULT_REGION;
    await dbConn.execute(`SET s3_region='${s3_region}'`);
    await dbConn.execute(`SET s3_access_key_id='${aws_access_key_id}'`);
    await dbConn.execute(
      `SET s3_secret_access_key='${process.env.AWS_SECRET_ACCESS_KEY}'`
    );
    log.debug("initS3: done.");
  }
}
