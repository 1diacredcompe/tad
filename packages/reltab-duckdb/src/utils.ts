import { Connection } from "ac-node-duckdb";

export const dbAll = async (
  dbConn: Connection,
  query: string
): Promise<any> => {
  const resIter = await dbConn.executeIterator(query);
  const resRows = resIter.fetchAllRows();
  resIter.close();
  return resRows;
};
