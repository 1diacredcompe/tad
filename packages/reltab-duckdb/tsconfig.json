{
  "compilerOptions": {
    "outDir": "./dist",
    "declaration": true,
    "target": "ES5",
    "strict": true,
    "lib": ["es6", "es2017", "dom"]
  },
  "include": ["./src/*", "./test/*.ts"]
}
