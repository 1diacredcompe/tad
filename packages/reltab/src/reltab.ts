export * from "./dialectRegistry";
export * from "./Connection";
export * from "./defs";
export * from "./FilterExp";
export * from "./QueryExp";
export * from "./Schema";
export * from "./TableRep";
export * from "./AggFn";
export * from "./ColumnType";
export * from "./DataSource";
