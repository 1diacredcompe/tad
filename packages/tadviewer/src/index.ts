/*
export * from "./components/AggPanel";
export * from "./components/ColumnList";
export * from "./components/ColumnRow";
export * from "./components/ColumnSelector";
export * from "./components/DisplayOrderPanel";
export * from "./components/FilterEditor";
export * from "./components/FilterEditorRow";
export * from "./components/Footer";
export * from "./components/FormatPanel";
export * from "./components/GridPane";
export * from "./components/IndeterminateCheckbox";
export * from "./components/LoadingModal";
export * from "./components/NumFormatPanel";
export * from "./components/PivotOrderPanel";
export * from "./components/Sidebar";
export * from "./components/SingleColumnSelect";
export * from "./components/SortOrderPanel";
export * from "./components/TextFormatPanel";
*/

export * from "./AppState";
export * from "./components/AppPane";
