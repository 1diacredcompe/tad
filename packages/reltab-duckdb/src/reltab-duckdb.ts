import * as tp from "typed-promisify";
import { Connection, DuckDB, IDuckDBConfig } from "node-duckdb";
import * as log from "loglevel";
import {
  TableRep,
  QueryExp,
  Schema,
  TableInfoMap,
  TableInfo,
  Row,
  ColumnMetaMap,
  DataSourceConnection,
  ColumnType,
  DataSourcePath,
  DataSourceNode,
  DataSourceId,
  EvalQueryOptions,
  DataSourceProvider,
  registerProvider,
  defaultEvalQueryOptions,
  DuckDBDialect,
} from "reltab"; // eslint-disable-line
import { SQLDialect } from "reltab/dist/dialect";
import { initS3 } from "./s3utils";
import { dbAll } from "./utils";

export * from "./csvimport";

const columnTypes = DuckDBDialect.columnTypes;

const typeLookup = (tnm: string): ColumnType => {
  const ret = columnTypes[tnm] as ColumnType | undefined;
  if (ret == null) {
    throw new Error("typeLookup: unknown type name: '" + tnm + "'");
  }
  return ret;
};

/* A little ConnectionPool class because it turns out node-duckdb
 * doesn't allow concurrent queries on a connection.
 */
class ConnectionPool {
  db: DuckDB;
  private pool: Connection[];

  constructor(db: DuckDB) {
    this.db = db;
    this.pool = [];
  }

  async take(): Promise<Connection> {
    if (this.pool.length > 0) {
      return this.pool.pop()!;
    } else {
      const conn = new Connection(this.db);
      await initS3(conn);
      return conn;
    }
  }

  giveBack(conn: Connection) {
    this.pool.push(conn);
  }
}

export class DuckDBContext implements DataSourceConnection {
  readonly displayName: string;
  readonly sourceId: DataSourceId;
  dbfile: string;
  db: DuckDB;
  connPool: ConnectionPool;
  private tableMap: TableInfoMap;

  constructor(dbfile: string, db: DuckDB, dbConn: Connection) {
    this.dbfile = dbfile;
    this.displayName = dbfile;
    this.sourceId = { providerName: "duckdb", resourceId: dbfile };
    this.db = db;
    this.connPool = new ConnectionPool(db);
    this.tableMap = {};
  }

  async runSQLQuery(query: string): Promise<any> {
    console.log("runSQLquery: ", query);
    const conn = await this.connPool.take();
    let ret: any;
    try {
      ret = await dbAll(conn, query);
    } finally {
      this.connPool.giveBack(conn);
    }
    return ret;
  }

  async getDisplayName(): Promise<string> {
    return this.displayName;
  }

  // ensure every table mentioned in query is registered:
  async ensureTables(query: QueryExp): Promise<void> {
    const tblNames = query.getTables();
    const namesArr = Array.from(tblNames);
    for (let tblName of namesArr) {
      if (this.tableMap[tblName] === undefined) {
        await this.getTableInfo(tblName);
      }
    }
  }

  async getSchema(query: QueryExp): Promise<Schema> {
    await this.ensureTables(query);
    const schema = query.getSchema(DuckDBDialect, this.tableMap);
    return schema;
  }

  async evalQuery(
    query: QueryExp,
    offset?: number,
    limit?: number,
    options?: EvalQueryOptions
  ): Promise<TableRep> {
    let t0 = process.hrtime();
    await this.ensureTables(query);
    const schema = query.getSchema(DuckDBDialect, this.tableMap);
    const sqlQuery = query.toSql(DuckDBDialect, this.tableMap, offset, limit);
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    const trueOptions = options ? options : defaultEvalQueryOptions;

    if (trueOptions.showQueries) {
      // log.info("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.info("DuckDBContext.evalQuery: evaluating:\n" + sqlQuery);
    }

    const t2 = process.hrtime();
    const dbRows = await this.runSQLQuery(sqlQuery);
    const rows = dbRows as Row[];
    const t3 = process.hrtime(t2);
    const [t3s, t3ns] = t3;
    const t4pre = process.hrtime();
    const ret = new TableRep(schema, rows);
    const t4 = process.hrtime(t4pre);
    const [t4s, t4ns] = t4;

    /*
    if (this.showQueries) {
      log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
      log.info("time to mk table rep: %ds %dms", t4s, t4ns / 1e6);
    }
    */

    return ret;
  }

  async rowCount(query: QueryExp, options?: EvalQueryOptions): Promise<number> {
    let t0 = process.hrtime();
    await this.ensureTables(query);
    const countSql = query.toCountSql(DuckDBDialect, this.tableMap);
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    const trueOptions = options ? options : defaultEvalQueryOptions;

    if (trueOptions.showQueries) {
      // log.info("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("DuckDBContext.rowCount: evaluating: \n" + countSql);
    }

    const t2 = process.hrtime();
    const qp = this.runSQLQuery(countSql);
    return qp.then((rows) => {
      const t3 = process.hrtime(t2);
      const [t3s, t3ns] = t3;
      /*
      if (this.showQueries) {
        log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
      }
      */
      const ret = Number.parseInt(rows[0].rowCount);
      return ret;
    });
  }

  dbGetTableInfo(tableName: string): Promise<TableInfo> {
    const tiQuery = `PRAGMA table_info(${tableName})`;
    console.log("*** dbGetTableInfo: tiQuery: ", tiQuery);
    const qp = this.runSQLQuery(tiQuery);
    return qp.then((dbRows) => {
      const rows = dbRows as Row[];
      console.log("getTableInfo: ", rows);
      log.debug("getTableInfo: ", rows);

      const extendCMap = (
        cmm: ColumnMetaMap,
        row: any,
        idx: number
      ): ColumnMetaMap => {
        const cnm = row.name;
        const cType = row.type.toLocaleUpperCase();

        if (cType == null) {
          log.error(
            'mkTableInfo: No column type for "' + cnm + '", index: ' + idx
          );
        }
        const cmd = {
          displayName: cnm,
          columnType: cType,
        };
        cmm[cnm] = cmd;
        return cmm;
      };

      const cmMap = rows.reduce(extendCMap, {});
      const columnIds = rows.map((r) => r.name);
      const schema = new Schema(DuckDBDialect, columnIds as string[], cmMap);
      return {
        tableName,
        schema,
      };
    });
  }

  async getTableInfo(tableName: string): Promise<TableInfo> {
    let ti = this.tableMap[tableName];
    if (!ti) {
      ti = await this.dbGetTableInfo(tableName);
      if (ti) {
        this.tableMap[tableName] = ti;
      }
    }
    return ti;
  }

  async getRootNode(): Promise<DataSourceNode> {
    const rootNode: DataSourceNode = {
      id: this.dbfile,
      kind: "Database",
      displayName: this.dbfile,
      isContainer: true,
    };
    return rootNode;
  }
  async getChildren(dsPath: DataSourcePath): Promise<DataSourceNode[]> {
    const { path } = dsPath;
    let node: DataSourceNode;
    const tiQuery = `PRAGMA show_tables;`;
    const dbRows = await this.runSQLQuery(tiQuery);
    const tableNames: string[] = dbRows.map((row: any) => row.name);
    const childNodes: DataSourceNode[] = tableNames.map((tableName) => ({
      id: tableName,
      kind: "Table",
      displayName: tableName,
      isContainer: false,
    }));
    return childNodes;
  }

  async getTableName(dsPath: DataSourcePath): Promise<string> {
    const { path } = dsPath;
    if (path.length < 1) {
      throw new Error("getTableName: empty path");
    }
    return path[path.length - 1];
  }
}

const duckdbDataSourceProvider: DataSourceProvider = {
  providerName: "duckdb",
  connect: async (resourceId: any): Promise<DataSourceConnection> => {
    const dbfile = resourceId as string;
    const dbOpts: IDuckDBConfig = {};
    if (dbfile) {
      dbOpts.path = dbfile;
    }
    const db = new DuckDB(dbOpts);
    const dbConn = new Connection(db);
    const ctx = new DuckDBContext(dbfile, db, dbConn);
    return ctx;
  },
};

registerProvider(duckdbDataSourceProvider);
