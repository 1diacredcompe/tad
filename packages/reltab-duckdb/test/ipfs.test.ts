import * as duckdb from "ac-node-duckdb";
import * as reltab from "reltab";
import * as reltabDuckDB from "../src/reltab-duckdb";
import * as tp from "typed-promisify";
import { textSpanContainsPosition, textSpanContainsTextSpan } from "typescript";
import { delimiter } from "path";
import * as log from "loglevel";
import * as util from "./testUtils";
import * as _ from "lodash";
import { asString, Row, Schema, tableQuery, TableRep } from "reltab";
import { getFormattedRows } from "./testUtils";

let testCtx: reltabDuckDB.DuckDBContext;

beforeAll(async (): Promise<reltabDuckDB.DuckDBContext> => {
  const ctx = await reltab.getConnection({
    providerName: "duckdb",
    resourceId: ":memory:",
  });

  testCtx = ctx as reltabDuckDB.DuckDBContext;

  return testCtx;
});

const importParquet = async (
  db: duckdb.DuckDB,
  path: string
): Promise<string> => {
  const tableName = await reltabDuckDB.nativeParquetImport(db, path);
  return tableName;
};

test("https import test", async () => {
  const tableName = await importParquet(
    testCtx.db,
    "https://github.com/deepcrawl/node-duckdb/raw/master/src/tests/test-fixtures/alltypes_plain.parquet"
  );

  // console.log("https import complete, tableName: ", tableName);

  const q1 = reltab.tableQuery(tableName);
  const qres = await testCtx.evalQuery(q1);
  // console.log("basic tableQuery result: ", qres);

  const q2 = q1.project([
    "id",
    "bool_col",
    "int_col",
    "string_col",
    "timestamp_col",
  ]);
  const q2res = await testCtx.evalQuery(q2);

  const q2sres = JSON.stringify(q2res, null, 2);

  //  console.log("project query result: ", q2sres );
  expect(q2sres).toMatchSnapshot();

  const fmtRows = getFormattedRows(qres);
  // console.log("fmtRows: ", fmtRows);
  expect(fmtRows).toMatchSnapshot();
});

test("s3 import test", async () => {
  const dbc = testCtx;

  let importSucceeded = false;
  let tableName: string = "";
  // const s3URL = 's3://ursa-labs-taxi-data/2009/01/data.parquet';
  const s3URL =
    "s3://amazon-reviews-pds/parquet/product_category=Books/part-00000-495c48e6-96d6-4650-aa65-3c36a3516ddd.c000.snappy.parquet";
  try {
    tableName = await importParquet(testCtx.db, s3URL);
    console.log("s3 import complete, tableName: ", tableName);
    importSucceeded = true;
  } catch (err) {
    console.error("caught error during s3 import: ", err);
  }
  expect(importSucceeded).toBe(true);
}, 15000);
