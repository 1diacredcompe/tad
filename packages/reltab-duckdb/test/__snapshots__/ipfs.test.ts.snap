// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`https import test 1`] = `
"{
  \\"schema\\": {
    \\"dialect\\": \\"duckdb\\",
    \\"columns\\": [
      \\"id\\",
      \\"bool_col\\",
      \\"int_col\\",
      \\"string_col\\",
      \\"timestamp_col\\"
    ],
    \\"columnMetadata\\": {
      \\"id\\": {
        \\"displayName\\": \\"id\\",
        \\"columnType\\": \\"INTEGER\\"
      },
      \\"bool_col\\": {
        \\"displayName\\": \\"bool_col\\",
        \\"columnType\\": \\"BOOLEAN\\"
      },
      \\"int_col\\": {
        \\"displayName\\": \\"int_col\\",
        \\"columnType\\": \\"INTEGER\\"
      },
      \\"string_col\\": {
        \\"displayName\\": \\"string_col\\",
        \\"columnType\\": \\"BLOB\\"
      },
      \\"timestamp_col\\": {
        \\"displayName\\": \\"timestamp_col\\",
        \\"columnType\\": \\"TIMESTAMP\\"
      }
    }
  },
  \\"rowData\\": [
    {
      \\"id\\": 4,
      \\"bool_col\\": true,
      \\"int_col\\": 0,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          48
        ]
      },
      \\"timestamp_col\\": 1235865600000
    },
    {
      \\"id\\": 5,
      \\"bool_col\\": false,
      \\"int_col\\": 1,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          49
        ]
      },
      \\"timestamp_col\\": 1235865660000
    },
    {
      \\"id\\": 6,
      \\"bool_col\\": true,
      \\"int_col\\": 0,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          48
        ]
      },
      \\"timestamp_col\\": 1238544000000
    },
    {
      \\"id\\": 7,
      \\"bool_col\\": false,
      \\"int_col\\": 1,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          49
        ]
      },
      \\"timestamp_col\\": 1238544060000
    },
    {
      \\"id\\": 2,
      \\"bool_col\\": true,
      \\"int_col\\": 0,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          48
        ]
      },
      \\"timestamp_col\\": 1233446400000
    },
    {
      \\"id\\": 3,
      \\"bool_col\\": false,
      \\"int_col\\": 1,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          49
        ]
      },
      \\"timestamp_col\\": 1233446460000
    },
    {
      \\"id\\": 0,
      \\"bool_col\\": true,
      \\"int_col\\": 0,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          48
        ]
      },
      \\"timestamp_col\\": 1230768000000
    },
    {
      \\"id\\": 1,
      \\"bool_col\\": false,
      \\"int_col\\": 1,
      \\"string_col\\": {
        \\"type\\": \\"Buffer\\",
        \\"data\\": [
          49
        ]
      },
      \\"timestamp_col\\": 1230768060000
    }
  ]
}"
`;

exports[`https import test 2`] = `
Array [
  Array [
    "4",
    "true",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "03/01/09",
    "0",
    "2009-03-01T00:00:00.000Z",
  ],
  Array [
    "5",
    "false",
    "1",
    "1",
    "1",
    "10",
    "1.100000023841858",
    "10.1",
    "03/01/09",
    "1",
    "2009-03-01T00:01:00.000Z",
  ],
  Array [
    "6",
    "true",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "04/01/09",
    "0",
    "2009-04-01T00:00:00.000Z",
  ],
  Array [
    "7",
    "false",
    "1",
    "1",
    "1",
    "10",
    "1.100000023841858",
    "10.1",
    "04/01/09",
    "1",
    "2009-04-01T00:01:00.000Z",
  ],
  Array [
    "2",
    "true",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "02/01/09",
    "0",
    "2009-02-01T00:00:00.000Z",
  ],
  Array [
    "3",
    "false",
    "1",
    "1",
    "1",
    "10",
    "1.100000023841858",
    "10.1",
    "02/01/09",
    "1",
    "2009-02-01T00:01:00.000Z",
  ],
  Array [
    "0",
    "true",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "01/01/09",
    "0",
    "2009-01-01T00:00:00.000Z",
  ],
  Array [
    "1",
    "false",
    "1",
    "1",
    "1",
    "10",
    "1.100000023841858",
    "10.1",
    "01/01/09",
    "1",
    "2009-01-01T00:01:00.000Z",
  ],
]
`;
